torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -273,14 +254,33 @@ void ProcessGroupAgent::sync() {
 }
 
 void ProcessGroupAgent::start() {
+  rpcRunning_.store(true);
   listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
   futureTimeoutThread_ =
       std::thread(&ProcessGroupAgent::pollTimedOutRPCs, this);
 }
 
+void ProcessGroupAgent::shutdown() {
+  if (!rpcRunning_.exchange(false)) {
+    return;
+  }
+  LOG(INFO) << "Stopping ProcessGroupAgent.";
+  {
+    std::unique_lock<std::mutex> lock(recvWorkMutex_);
+    if (recvWork_) {
+      recvWork_->abort();
+    }
+  }
+  threadPool_.waitWorkComplete();
+  listenerThread_.join();
+  futureTimeoutCV_.notify_one();
```

Comment:
What if the `futureTimeoutThread_` is not blocked on the `futureTimeoutCV_`? For example, it could be at line `std::chrono::milliseconds sleepTime;`.

Will `futureTimeoutThread_` live forever?

Reply:
Thanks for pointing this out - so it sounds like this should be fixed if we hold the lock on `rpcRunning`? Should we hold that lock wherever we write/read it?